### PR TITLE
Use declaritive system for creating singletons

### PR DESCRIPTION
**Pull request by Nokel81**
- Let the code say the requirements of each singleton is instead of
  comments

Signed-off-by: Sebastian Malton <sebastian@malton.name>

### CODE REVIEW

**Comment by jakolehm** [comment]
If this does not fix a bug I would move it to next minor version.

File: src/common/utils/singleton.ts [hunk 1]
```diff
@@ -88,3 +88,94 @@ export class Singleton {
...
+
+interface CreateRequirements<T extends typeof Singleton> {
+  builds: T,
+  args?: ConstructorParameters<T> | (() => ConstructorParameters<T>),
+  requires?: Iterable<typeof Singleton>,
+}
+
+interface DeclareOpts {
+  requires?: Iterable<typeof Singleton>,
+
```
Comment (panuhorsmalahti) [inline src/common/utils/singleton.ts:98]: [nit] extra empty line

File: src/common/utils/singleton.ts [hunk 2]
```diff
@@ -88,3 +88,94 @@ export class Singleton {
...
+}
+
+/**
+ * A builder class for declaring the dependency graph of a set of singletons
+ *
+ * This is useful because it allows the code to declare the requirements
+ */
+export class CreateSingletons {
+  rawRequirements: CreateRequirements<typeof Singleton>[] = [];
```
Comment (panuhorsmalahti) [inline src/common/utils/singleton.ts]: This should probably be marked as private?

File: src/renderer/bootstrap.tsx [hunk 3]
```diff
@@ -84,40 +84,42 @@ export async function bootstrap(App: AppComponent) {
...
 
   // Register additional store listeners
-  cs.registerIpcListener();
+  ClusterStore.getInstance().registerIpcListener();
 
   // init app's dependencies if any
-  if (App.init) {
-    await App.init();
-  }
+  await App.init?.();
```
Comment (panuhorsmalahti) [inline src/renderer/bootstrap.tsx]: I don't think change improves readability

**Review by panuhorsmalahti — DISMISSED** [review]

File: src/common/utils/singleton.ts (hunk 2 above)
Comment (Nokel81) [reply src/common/utils/singleton.ts]: Yes

File: src/renderer/bootstrap.tsx (hunk 3 above)
Comment (Nokel81) [reply src/renderer/bootstrap.tsx]: Okay I'll revert.

**Comment by Nokel81** [comment]
Closing because we will probably be moving to DI

**Pull request closed without merging**